cachi2/core/package_managers/npm.py

```diff
@@ -173,6 +185,199 @@ def get_sbom_components(self) -> list[dict[str, str]]:
         return [{"name": package.name, "version": package.version} for package in packages]
 
+    def get_dependencies_to_download(self) -> Dict[str, Dict[str, Optional[str]]]:
+        """Return a Dict of URL dependencies to download."""
+        packages = self._dependencies if self.lockfile_version == 1 else self._packages
+        return {
+            package.name: {
+                "version": package.version,
+                "url": package.resolved_url,
+                "integrity": package.integrity,
+            }
+            for package in packages
+            if "file:" not in package.resolved_url
+        }
+
+
+def _update_vcs_url_with_full_hostname(vcs: str) -> str:
+    """Update VCS URL with full hostname.
+
+    Transform github:kevva/is-positive#97edff6
+    into git+ssh://github.com/kevva/is-positive.git#97edff6
+
+    :param vcs: VCS URL to be modified with full hostname and file extension
+    :return: Updated VCS URL
+    """
+    host, _, path = vcs.partition(":")
+    namespace_repo, _, ref = path.partition("#")
+    suffix_domain = "org" if host == "bitbucket" else "com"
+
+    vcs = f"git+ssh://{host}.{suffix_domain}/{namespace_repo}.git"
+    if ref:
+        vcs = f"{vcs}#{ref}"
+    return vcs
+
+
+def _extract_git_info_npm(vcs_url: str) -> Dict[str, str]:
+    """
+    Extract important info from a VCS requirement URL.
+
+    Given a URL such as git+ssh://user@host/namespace/repo.git#9e164b970
+
+    this function will extract:
+    - the "clean" URL: ssh://user@host/namespace/repo.git
+    - the git ref: 9e164b970
+
+    The clean URL and ref can be passed straight to scm.Git to fetch the repo.
+    The host, namespace and repo will be used to construct the file path under deps/npm.
+
+    :param vcs_url: The URL of a VCS requirement, must be valid (have git ref in path)
+    :return: Dict with url, ref, host, namespace and repo keys
+    """
+    # If scheme is git+protocol://, keep only protocol://
+    # Do this before parsing URL, otherwise urllib may not extract URL params
+    if vcs_url.startswith("git+"):
+        vcs_url = vcs_url[len("git+") :]
+
+    clean_url, _, ref = vcs_url.partition("#")
+
+    url = urlparse(clean_url)
+    namespace_repo = url.path.strip("/")
+    if namespace_repo.endswith(".git"):
+        namespace_repo = namespace_repo[: -len(".git")]
+
+    # Everything up to the last '/' is namespace, the rest is repo
+    namespace, _, repo = namespace_repo.partition("/")
+
+    vcs_url_info = {
+        "url": clean_url,
+        "ref": ref.lower(),
+        "namespace": namespace,
+        "repo": repo,
+    }
+
+    for key, value in vcs_url_info.items():
+        if not value:
+            raise FetchError(f"{vcs_url} is not valid VCS url. {key} is missing.")
+
+    if url.hostname:
+        vcs_url_info["host"] = url.hostname
+    else:
+        raise FetchError(f"{vcs_url} is not valid VCS url. Host is missing.")
```

Comment:
Same here, PackageRejected or UnexpectedFormat (I'd go with unexpected format, if this error occurs it's probably because we missed some corner case)

Note that the default "solution" for FetchError suggests retrying, which definitely won't help here :smile:

Reply:
Makes sense I took this phase as fetching deps hence the FetchError. :D I am fixing those